src/test/java/org/jenkinsci/plugins/workflow/graph/FlowNodeSerialWalkerTest.java

```diff
@@ -0,0 +1,114 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.graph;
+
+import com.google.inject.Inject;
+import hudson.model.TaskListener;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepDescriptorImpl;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepImpl;
+import org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+import org.junit.Test;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class FlowNodeSerialWalkerTest {
+
+    @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule public JenkinsRule r = new JenkinsRule();
+
+    @Test public void labelEnclosing() throws Exception {
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(
+            "lE 'start'; parallel one: {lE 'in-one'}, two: {lE 'in-two'}; lE 'middle'\n" +
+            "stage 'dev'; lE 'in-dev'; stage 'test'; lE 'in-test'\n" +
+            "parallel three: {lE 'in-three'}\n" +
+            "node {writeFile file: 'f1', text: ''; load 'f1'; lE 'after-load'}\n" +
+            "node {writeFile file: 'f2', text: 'lE \"in-load\"'; load 'f2'}\n" +
+            "node {writeFile file: 'f3', text: '{-> lE \"closure-load\"}'; load('f3')()}\n" +
+            "lE 'end'", true));
+        WorkflowRun b = r.assertBuildStatusSuccess(p.scheduleBuild2(0));
+        r.assertLogContains("start: null", b);
+        r.assertLogContains("in-one: Branch: one", b);
+        r.assertLogContains("in-two: Branch: two", b);
+        r.assertLogContains("middle: null", b);
+        r.assertLogContains("in-dev: dev", b);
+        r.assertLogContains("in-test: test", b);
+        r.assertLogContains("in-three: Branch: three", b);
+        r.assertLogContains("after-load: test", b);
```

Comment:
Yes it is a parent according to the graph, but it does not enclose the `lE` step. Specifically `currentLabel()` will be null on that iteration since `isAncestor()` will be false: the iterator would first have hit the `BlockEndNode` of the `load` step, omitting it in favor of the corresponding `BlockStartNode` (`StepStartNode` to be precise), and setting `ancestor = false`.